node/building_blocks.re

```diff
@@ -55,6 +55,29 @@ let is_current_producer = (state, ~key_hash) => {
   Some(current_producer.address == key_hash);
 };
 
+// TODO: verify reasonable times for this with benchmarking.
+// currently set to [state_root_min_timeout] as a reasonable
+// default.
+let minimum_signable_time_between_epochs = 60.0;
+
+// TODO: verify reasonable times for this with benchmarking.
+// Currently set to [minimum_signable_time_between_epochs] + 10
+// as a reasonable default.
+let maximum_signable_time_between_epochs = 70.0;
+
+let block_has_signable_state_root_hash = (~current_time, state, block) => {
+  let protocol = state.Node.protocol;
+  let time_since_last_epoch = current_time -. protocol.last_state_root_update;
+  Crypto.(
+    if (BLAKE2B.equal(block.Block.state_root_hash, protocol.state_root_hash)) {
+      time_since_last_epoch <= maximum_signable_time_between_epochs;
```

Comment:
What exactly are these two conditions and how/why do they guarantee a signable state root hash? As far as I can tell, looks like here we're saying:

If a block's state root hash is the same as the state root hash published by Tezos (? assuming that's what `protocol.state_root_hash` is), we have a signable state root hash if the time since the last epoch is less than or equal to the maximum.

Otherwise, that is, if the block's state root hash and the published state root hash differ, we have a signable state root hash if the block's state root hash is equal to the next state root hash held by the state, and we've exceeded the time since the last epoch.

This seems to reduce to "If we're consistent with the published SRH and aren't late to the next epoch, we may sign the SRH. Otherwise, if we're consistent with the next SRH and have exceeded the time allotted to the last epoch, we may sign the SRH.

Is this correct? Is my use of language or my comprehension faulty?

Reply:
What I'm trying to say is:

You may sign the block iff
- Either, the SRH of the block matches the current SRH you have AND the epoch change is not late (i.e, the time between now and the last state root update is less than the `maximum_signable_time_between_epochs`)
- Or, the SRH matches the `next_state_root_hash` you calculated previously, and the epoch change is not too early (i.e the time between now and the last state root update is greater than the `minimum_signable_time_between_blocks`)

Does that make sense?

Reply:
Totally, thanks